### ISSUE TITLE
LongPolling Bot stops receiving updates after exactly 20 requests

### ISSUE THREAD

**Issue by pouraghajan** [rubenlagus/TelegramBots#448]
Hi,
I'm using library version 3.6.1 and testing a LongPolling bot in my development ENV.
I should emphasis that, I have already done this from A to Z couple of times before and never faced this issue.
My problem is, the bot starts and exactly after serving 20 requests (means receiving 20 updates, regardless of the time span during which it receives them) it stops receiving more. There is no track of any exceptions or errors what so ever. I'm sure this is not because I have sent more than 20 requests to my bot in one minute (which is Telegram bot API's limitation) because, firstly I have simulated the case in a way that 20 updates would be sent to my bot in more than 1 minute, and secondly, if that would be the case, the bot (or my Telegram account in this case) should be unblocked after a while, which is not happening in my case.
I create and register my bot as following (which is exactly what its mentioned in this repository's example codes). I have even placed a log command in the first line of my onUpdateReceived() method, in order to make sure I would be notified upon receiving any new update.
But it seems the java Thread which is being instantiated with my bot session during the bot registration process is being killed after receiving 20 messages and it never recovers. So the updates still are coming toward my endpoint but there is no UpdateHandler to serve them (I'm sure about it because I have used WireShark to monitor the network traffic, and I'm confident the update requests are coming from Telegram server to my development machine).
I know there should be some neglects in my own code and this can't be an internal library issue, but I would appreciate any help, leading me to the root cause. The only difference I could determine between my code and example bots code is, I have used other classes (Model like classes) to implement the logic behind my bot's features and instantiated them in a normal way like any other object but the service classes and so many other methods in library examples are being instantiated and called in "static" and "synchronized" fashion. I'm not even sure if this could be relevant.
 (Frustrated like sh...t!!!)

```
public class MainBot {
    public static void main(String[] args) {
        BotLogger.setLevel(Level.ALL);
        BotLogger.registerLogger(new ConsoleHandler());
        try {
            BotLogger.registerLogger(new BotsFileHandler());
        } catch (IOException e) {
            BotLogger.severe(LOGTAG, e);
        }
        try {
            ApiContextInitializer.init();
            TelegramBotsApi telegramBotsApi = createTelegramBotsApi();
            try {
                // Register long polling bots. They work regardless type of TelegramBotsApi we are creating
                MyLPBot longPollingBot = new MyLPBot();
                telegramBotsApi.registerBot(longPollingBot);
            } catch (TelegramApiException e) {
                BotLogger.error(LOGTAG, e);
            }
        } catch (Exception e) {
            BotLogger.error(LOGTAG, e);
        }
    }

    private static TelegramBotsApi createTelegramBotsApi() throws TelegramApiException {
        TelegramBotsApi telegramBotsApi = createLongPollingTelegramBotsApi();
        return telegramBotsApi;
    }

    /**
     * @brief Creates a Telegram Bots Api to use Long Polling (getUpdates) bots.
     * @return TelegramBotsApi to register the bots.
     */
    private static TelegramBotsApi createLongPollingTelegramBotsApi() {
        return new TelegramBotsApi();
    }
}
```

**Comment by rubenlagus** [comment]
@pouraghajan Can you try this replacing the myLPBot with just an echo one? A bot that will just answer any message with its own content (similar to the one in the starting guide.

This may help to isolate the issue.

**Comment by pouraghajan** [comment]
Hi,
I've tested what you suggest and yes, when I only echo the update message, there is no problem anymore and I can receive more than 20 updates.
So as I assumed, the issue should be somewhere in my background Bot logic classes.
Let me explain that part of my code, so maybe we can pinpoint the problem.

I have a set of Model (or service) classes to do the logical process behind the scene per each update message received from user. I track user's state among different pages of the bot and call one of these classes based on the current state user has. These classes are being instantiated in MyLPBot class inside the "onUpdateReceived()" method. Here is one example of my code:

```
public void onUpdateReceived(Update update) {
        int sessionTid = 0;
        String chatId = new String();
        if (update.hasMessage()) {
            Message message = update.getMessage();
            sessionTid = message.getFrom().getId();
            chatId = String.valueOf(message.getChatId());
        }
        if (update.hasCallbackQuery()) {
            CallbackQuery callbackQuery = update.getCallbackQuery();
            sessionTid = callbackQuery.getFrom().getId();
            chatId = String.valueOf(callbackQuery.getMessage().getChatId());
        }
            daoMain = new DaoMain();
            Session session = daoMain.fetchSession(sessionTid, chatId);
switch (session.getState())
                {
                    case BotStates.MAIN_MENU :
                    {
                        MainMenu mainMenu = new MainMenu(session, update);
                        OutputMessage outputMessage = mainMenu.process();
                        sendResponse(outputMessage);
                        break;
                    }
               }
}
```

Then in the Model class (MainMenu in this example) I process the update received from user (considering the state it has in the Bot life-cycle which is in the "session" object) and then create and send back a bean object called OutputMessage (which looks like as following) and finally call the "sendResponse" method in MyLPBot class to send back the bot's response to user.

```
public class OutputMessage {
    private String chatId;
    private Object messageObject;

    public OutputMessage(String chatId, Object messageObject) {
        this.chatId= chatId;
        this.messageObject = messageObject;
    }

    public String getchatId() {
        return chatId;
    }

    public void setUserId(String chatId) {
        this.chatId= chatId;
    }

    public Object getMessageObject() {
        return messageObject;
    }
}
```

```
private Message sendResponse(OutputMessage outputMessage)  {
        Message message = new Message();
        if (outputMessage.getMessageObject() instanceof SendMessage) {
            SendMessage sendMessage = (SendMessage) outputMessage.getMessageObject();
            try {
                message = execute(sendMessage);
            } catch (TelegramApiException e) {
                e.printStackTrace();
            }
        }
        else if (outputMessage.getMessageObject() instanceof SendVideo) {
            SendVideo sendVideo = (SendVideo) outputMessage.getMessageObject();
            try {
                message = sendVideo(sendVideo);
            } catch (TelegramApiException e) {
                e.printStackTrace();
            }
        }
        else if (outputMessage.getMessageObject() instanceof SendVideoNote) {
            SendVideoNote sendVideoNote = (SendVideoNote) outputMessage.getMessageObject();
            try {
                message = sendVideoNote(sendVideoNote);
            } catch (TelegramApiException e) {
                e.printStackTrace();
            }
        }
        else if (outputMessage.getMessageObject() instanceof SendPhoto) {
            SendPhoto sendPhoto = (SendPhoto) outputMessage.getMessageObject();
            try {
                message = sendPhoto(sendPhoto);
            } catch (TelegramApiException e) {
                e.printStackTrace();
            }
        }
}
```

**Comment by pouraghajan** [comment]
To whom, who may experienced the same issue as me:
The problem wasn't in this library but it was related to the Database Connection Pool management in my code.
I accidentally opened connections from my C3P0 connection pool while there was a resource leakage in connection handling (I wasn't closing all of them). So after opening 20 connections, the pool reaches to its maximum and all the new updates have to wait in the queue for acquiring new database connection.
Sorry if I misled somebody by opening this ticket.
